src/api.rs

```diff
@@ -0,0 +1,243 @@
+    if let Some(group) = new_hub.groups.get_mut(&new_hub.default_group) {
+        group.set_channel_permission(
+            channel_id.clone(),
+            crate::permission::ChannelPermission::ViewChannel,
+            crate::permission::PermissionSetting::TRUE,
+        );
+        group.set_channel_permission(
+            channel_id.clone(),
+            crate::permission::ChannelPermission::SendMessage,
+            crate::permission::PermissionSetting::TRUE,
+        );
+        group.set_channel_permission(
+            channel_id.clone(),
+            crate::permission::ChannelPermission::ReadMessage,
+            crate::permission::PermissionSetting::TRUE,
+        );
+    }
+    owner.in_hubs.push(id.clone());
+    owner.save().await?;
+    new_hub.save().await?;
+    Ok(id)
+}
+
+pub async fn get_hub(user: &User, hub_id: &ID) -> Result<Hub> {
+    user.in_hub(&hub_id)?;
+    let hub = Hub::load(hub_id).await?;
+    hub.strip(&user.id)
+}
+
+pub async fn delete_hub(user: &User, hub_id: &ID) -> Result<()> {
+    user.in_hub(&hub_id)?;
+    let hub = Hub::load(hub_id).await?;
+    let member = hub.get_member(&user.id)?;
+    if member.has_all_permissions() {
+        if let Ok(()) = tokio::fs::remove_file(hub.get_info_path()).await {
+            if let Ok(()) = tokio::fs::remove_dir_all(hub.get_data_path()).await {
+                Ok(())
+            } else {
+                Err(Error::DeleteFailed)
+            }
+        } else {
+            Err(Error::DeleteFailed)
+        }
+    } else {
+        Err(Error::NoPermission)
+    }
+}
+
+pub async fn rename_hub(user: &User, hub_id: &ID, new_name: String) -> Result<String> {
+    is_valid_username(&new_name)?;
+    user.in_hub(&hub_id)?;
+    let mut hub = Hub::load(hub_id).await?;
+    let member = hub.get_member(&user.id)?;
+    if member.has_permission(HubPermission::Administrate, &hub) {
+        let old_name = hub.name.clone();
+        hub.name = new_name;
+        hub.save().await?;
+        Ok(old_name)
+    } else {
+        Err(Error::NoPermission)
+    }
+}
+
+pub async fn change_nickname(user: &User, hub_id: &ID, new_name: String) -> Result<String> {
+    is_valid_username(&new_name)?;
+    user.in_hub(&hub_id)?;
+    let mut hub = Hub::load(hub_id).await?;
+    let member = hub.get_member_mut(&user.id)?;
+    let old_name = member.nickname.clone();
+    member.nickname = new_name;
+    hub.save().await?;
+    Ok(old_name)
+}
+
+pub async fn user_banned(user: &User, hub_id: &ID, user_id: &ID) -> Result<bool> {
+    user.in_hub(hub_id)?;
+    let hub = Hub::load(hub_id).await?;
+    Ok(hub.bans.contains(user_id))
+}
+
+pub async fn user_muted(user: &User, hub_id: &ID, user_id: &ID) -> Result<bool> {
+    user.in_hub(hub_id)?;
+    let hub = Hub::load(hub_id).await?;
+    Ok(hub.mutes.contains(user_id))
+}
+
+pub async fn get_hub_member(user: &User, hub_id: &ID, user_id: &ID) -> Result<HubMember> {
+    user.in_hub(hub_id)?;
```

Comment:
I think some of these constraints would be really nice and cool if you write your own attribute-like macros. Some of these checks can just be an attribute: `#[in_hub]` which might look nicer, but you might have trouble having a reference to the hub in the transformed function, which you might define a parameter in the function declaration but your macro removes it and set it as a local variable.

Or you can define your `macro_rules` macro which might be easier, with a syntax similar to this: 
```rust
in_hub!(hub_id, { hub => 
 hub.do_something();
})
```

Reply:
That would be really nice actually, also, then I would learn how to write attribute macros...

Reply:
I will most likely do this in a separate PR.